src/vm/methoddescbackpatchinfo.cpp

```diff
@@ -0,0 +1,241 @@
+
+    m_dependentLoaderAllocatorsWithSlotsToBackpatch->Remove(dependentLoaderAllocator);
+}
+
+#endif // !DACCESS_COMPILE
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// MethodDescBackpatchInfoTracker
+
+CrstStatic MethodDescBackpatchInfoTracker::s_lock;
+
+#ifndef DACCESS_COMPILE
+
+void MethodDescBackpatchInfoTracker::StaticInitialize()
+{
+    WRAPPER_NO_CONTRACT;
+    s_lock.Init(CrstMethodDescBackpatchInfoTracker);
+}
+
+#endif // DACCESS_COMPILE
+
+#ifdef _DEBUG
+
+bool MethodDescBackpatchInfoTracker::IsLockedByCurrentThread()
+{
+    WRAPPER_NO_CONTRACT;
+
+#ifndef DACCESS_COMPILE
+    return !!s_lock.OwnedByCurrentThread();
+#else
+    return true;
+#endif
+}
+
+bool MethodDescBackpatchInfoTracker::IsEligibleForEntryPointSlotBackpatch(PTR_MethodDesc methodDesc)
+{
+    // The only purpose of this method is to allow asserts in inline functions defined in the .h file, by which time MethodDesc
+    // is not fully defined
+
+    WRAPPER_NO_CONTRACT;
+    return methodDesc->IsEligibleForEntryPointSlotBackpatch();
+}
+
+#endif // _DEBUG
+
+#ifndef DACCESS_COMPILE
+
+MethodDescBackpatchInfo *MethodDescBackpatchInfoTracker::AddBackpatchInfo_Locked(MethodDesc *methodDesc)
+{
+    WRAPPER_NO_CONTRACT;
+    _ASSERTE(IsLockedByCurrentThread());
+    _ASSERTE(methodDesc != nullptr);
+    _ASSERTE(methodDesc->IsEligibleForEntryPointSlotBackpatch());
+    _ASSERTE(m_backpatchInfoHash.Lookup(methodDesc) == nullptr);
+
+    NewHolder<MethodDescBackpatchInfo> backpatchInfoHolder = new MethodDescBackpatchInfo(methodDesc);
+    m_backpatchInfoHash.Add(backpatchInfoHolder);
+    return backpatchInfoHolder.Extract();
+}
+
+EntryPointSlotsToBackpatch *MethodDescBackpatchInfoTracker::GetDependencyMethodDescEntryPointSlotsToBackpatch_Locked(
+    MethodDesc *methodDesc)
+{
+    WRAPPER_NO_CONTRACT;
+    _ASSERTE(IsLockedByCurrentThread());
+    _ASSERTE(methodDesc != nullptr);
+    _ASSERTE(methodDesc->IsEligibleForEntryPointSlotBackpatch());
+
+    MethodDescEntryPointSlotsToBackpatch *methodDescSlots =
+        m_dependencyMethodDescEntryPointSlotsToBackpatchHash.Lookup(methodDesc);
+    return methodDescSlots == nullptr ? nullptr : methodDescSlots->GetSlots();
+}
+
+EntryPointSlotsToBackpatch *MethodDescBackpatchInfoTracker::GetOrAddDependencyMethodDescEntryPointSlotsToBackpatch_Locked(
+    MethodDesc *methodDesc)
+{
+    WRAPPER_NO_CONTRACT;
+    _ASSERTE(IsLockedByCurrentThread());
+    _ASSERTE(methodDesc != nullptr);
+    _ASSERTE(methodDesc->IsEligibleForEntryPointSlotBackpatch());
+
+    MethodDescEntryPointSlotsToBackpatch *methodDescSlots =
+        m_dependencyMethodDescEntryPointSlotsToBackpatchHash.Lookup(methodDesc);
+    if (methodDescSlots != nullptr)
+    {
+        return methodDescSlots->GetSlots();
+    }
+
+    NewHolder<MethodDescEntryPointSlotsToBackpatch> methodDescSlotsHolder =
+        new MethodDescEntryPointSlotsToBackpatch(methodDesc);
```

Comment:
I think it's better to throw. For an app that intends to work well in memory-restricted environments, it may well be OOM-tolerant. If it is not tolerant then it's likely to fail pretty soon anyway so it doesn't matter much. Throwing would allow a retry to succeed without penalty and I'd prefer to avoid locking into a slow system with no recovery unless necessary. Going through the prestub on every call can be slow enough that it could appear as a bug and not just a perf issue. If the prestub could attempt to record slots that failed to be recorded before, that could be a solution, but not all of the info to do that is available currently and it may not be sustainable longer-term.

Reply:
A partial solution may be to catch the OOM from recording a slot in DoBackpatch() and let it try again on the next call to the prestub, and let the OOM propagate for VSD cases.